naptime/src/main/scala/org/coursera/naptime/ari/engine/Utilities.scala

```diff
@@ -36,51 +36,94 @@ import scala.collection.JavaConverters._
-      _ <- getTypedDefinition(schema)
-        if dataElement.getName.toString == "definition"
-      typeName <- Option(dataElement.getParent.getValue).flatMap {
-        case dataMap: DataMap => Option(dataMap.getString("typeName"))
-        case _ => None
-      }
-    } yield {
-      typeName
-    }).getOrElse(dataElement.getName.toString)
-    val parentNames = Option(dataElement.getParent).map(replaceTypedDefinitionNames)
-    (parentNames.getOrElse(List.empty) :+ thisName).filterNot(_.isEmpty)
+  private val TYPED_DEFINITION_KEY = "typedDefinition"
+  private def getTypedDefinitionMappings(elements: Iterator[DataElement]): Map[String, String] = {
+    val typedDefinitionMappings = elements.flatMap { element =>
+      Option(element.getSchema.getProperties.get(TYPED_DEFINITION_KEY)).collect {
+        case definitions: java.util.Map[String@unchecked, String@unchecked] => definitions.asScala.toList
+      }.getOrElse(List.empty)
+    }
+    typedDefinitionMappings.toMap
   }
 
   /**
-    * Iterates through a data map and returns the value stored at a particular path in the data map,
-    * or None if the path isn't found (or the data is null)
-    *
-    * @param element top-level element to be updated
-    * @param schema data schema defining the fields on the element
-    * @param path list of strings defining the path to the target element
-    */
-  private[naptime] def getValueAtPath(
+   * Iterates through a data map and returns the value(s) stored at a particular path in the data map,
+   * respecting `typedDefinition` annotations.
+   *
+   * We only iterate on value types to avoid needing to think about container and value types.
+   *
+   * This is a 1 to many operation: a single path may return multiple values.
+   * For example, the path is for an array of records: suppose
+   * the DataMap is {"courses": [{"id": "courseId1", "name": "my fav"}, {"id": "courseId2", "name": "not my fav"}]},
+   * then the path `courses/id` should return `["courseId1", "courseId2"]`.
+   *
+   * Another common case is if the path is for an array: suppose the DataMap is
+   * {"courseIds": ["courseId1", "courseId2"]}, then the path `courseIds` should return
+   * ["courseId1", "courseId2"]
+   *
+   * @param element top-level element for which we are retrieving the values.
+   * @param schema data schema defining the fields on the top-level element
+   * @param path list of strings defining the path to one or many target values.
+   */
+  private[naptime] def getValuesAtPath(
       element: DataMap,
       schema: RecordDataSchema,
-      path: Seq[String]): Option[Object] = {
-    val it = Builder.create(element, schema, IterationOrder.PRE_ORDER).dataIterator()
-    Iterator
-      .continually(it.next)
+      path: Seq[String]): List[String] = {
+
+    def getIterator() = {
+      val dataIterator = Builder.create(element, schema, IterationOrder.PRE_ORDER).dataIterator()
+
+      Iterator
+      .continually(dataIterator.next)
```

Comment:
nit: indentation/style; maybe we should run scalafmt on all naptime projects

Reply:
Yep, since we don't use Arcanist, I propose we use https://github.com/lucidsoftware/neo-sbt-scalafmt to autoformat on compile. I'll do that in a separate diff to not pollute this one.